[CIR] Integral types; simple global variables

dkolsen-pgi opened the pull request:
Add integral types to ClangIR.  These are the first ClangIR types, so the change includes some infrastructure for managing ClangIR types.

So that the integral types can be used somewhere, generate ClangIR for global variables using the new `cir.global` op.  As with the current support for functions, global variables are just a stub at the moment. The only properties that global variables have are a name and a type.

Add a new ClangIR code gen test global-var-simple.cpp, which defines global variables with most of the integral types.

(Part of upstreaming the ClangIR incubator project into LLVM.)

dkolsen-pgi committed:
[CIR] Integral types; simple global variables

Add integral types to ClangIR.  These are the first ClangIR types, so
the change includes some infrastructure for managing ClangIR types.

So that the integral types can be used somewhere, generate ClangIR for
global variables using the new cir.global op.  As with the current
support for functions, global variables are just a stub at the moment.
The only properties that global variables have are a name and a type.

Add a new ClangIR code gen test global-var-simple.cpp, which defines
global variables with most of the integral types.

(Part of upstreaming the ClangIR incubator project into LLVM.)

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -74,6 +75,32 @@ class LLVMLoweringInfo {
+def GlobalOp : CIR_Op<"global"> {
+  let summary = "Declare or define a global variable";
+  let description = [{
+    ... lots of text to be added later ...
```
Would still prefer a bit of a stub here, even if it is incomplete (rather than just a 'todo') as this makes it to our documentation.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -67,7 +67,8 @@ void CIRGenModule::emitGlobal(clang::GlobalDecl gd) {
   } else {
-    errorNYI(global->getSourceRange(), "global variable declaration");
+    const auto *vd = cast<VarDecl>(global);
+    assert(vd->isFileVarDecl() && "Cannot emit local var decl as global");
```
This is going to result in an 'unused variable' warning in release mode.  So inline hte 'cast' into the assert, or put a `(void)vd` at one point.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -82,6 +83,14 @@ void CIRGenModule::emitGlobalFunctionDefinition(clang::GlobalDecl gd,
+                                           bool isTentative) {
+  mlir::Type type = getTypes().convertType(vd->getType());
+  auto varOp = builder.create<cir::GlobalOp>(
+      getLoc(vd->getSourceRange()), vd->getIdentifier()->getName(), type);
```
`getIdentifier` could potentially result in a nullptr if this is a 'special' name (or with some other magic? I know it is possible for local variables).  So an assert/fallback for `getIdentifier==nullptr` would be appreciated here.

dkolsen-pgi replied:
Fixed.  The code checks `getIdentifier()`.  If it is null, an error is reported and the declaration is skipped.  This is temporary until the code handles non-identifier names.

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -0,0 +1,130 @@
+
+    Those integer types that are directly available in C/C++ standard are called
+    primitive integer types. Said types are: `signed char`, `short`, `int`,
+    `long`, `long long`, and their unsigned variations.
```
How is `_BitInt` represented?

erichkeane commented on clang/test/CIR/global-var-simple.cpp:
```diff
@@ -0,0 +1,53 @@
+char16_t c16;
+// CHECK: cir.global @c16 : !cir.int<u, 16>
+
+char32_t c32;
```
Can you do the saturated types as well?  I see them listed in your switch above.

dkolsen-pgi replied:
Listing the saturated types in the switch above was a mistake.  That has been fixed.  The saturated types have not been added to this test case.

dkolsen-pgi committed:
[CIR] Remove saturated types; add _BitInt types

This is a follow-up commit to "Integral types; simple global variables"

Remove the saturated integral types from the type conversion code, since
I am not confident that they are implemented correctly.  They are part
of fixed-point support, which has not yet been implemented in ClangIR.

Add _BitInt types to the type conversion code, and add test cases for
them.

Improve the documentation for FuncOp, GlobalOp, and IntType.

erichkeane commented on clang/test/CIR/global-var-simple.cpp:
```diff
@@ -0,0 +1,59 @@
+_BitInt(20) sb20;
+// CHECK: cir.global @sb20 : !cir.int<s, 20>
+
+unsigned _BitInt(48) ub48;
```
Would also like a test for 'absurdly large bitint' here, since they can be up to ~1.9 million bits in length.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenTypes.cpp:
```diff
@@ -0,0 +1,90 @@
+    case BuiltinType::Int:
+    case BuiltinType::Int128:
+    case BuiltinType::Long:
+    case BuiltinType::LongAccum:
```
I see some additional fixed-point types here (Accum/Fract).  Can we have tests for that, or let that fall-through and assert?

erichkeane reviewed (commented):
Can you point out the Github Issue for 'implement fixed-point' types?

AaronBallman commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -85,14 +115,15 @@ class CIR_Op<string mnemonic, list<Trait> traits = []> :
   let summary = "Declare or define a function";
   let description = [{
-    ... lots of text to be added later ...
+    THe `cir.func` operation defines a function, similar to the `mlir::FuncOp`
```
```suggestion
    The `cir.func` operation defines a function, similar to the `mlir::FuncOp`
```

dkolsen-pgi replied:
Fixed.

dkolsen-pgi commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -74,6 +75,32 @@ class LLVMLoweringInfo {
+  let skipDefaultBuilders = 1;
+
+  let builders = [OpBuilder<(ins "llvm::StringRef":$sym_name,
+		                             "mlir::Type":$sym_type)>];
```
This is new code.  It wasn't your fault.  The problem is that a couple tab characters made it in there, and VC Code expanded the tabs to a different width than GitHub.  I'll fix that.  Thanks for noticing that.

AaronBallman commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -0,0 +1,132 @@
+    /// Returns a minimum bitwidth of cir::IntType
+    static unsigned minBitwidth() { return 1; }
+    /// Returns a maximum bitwidth of cir::IntType
+    static unsigned maxBitwidth() { return 128; }
```
How does this handle `BITINT_MAXWIDTH` which is much, much wider than 128 on some targets?

dkolsen-pgi committed:
[CIR] Gracefully fail null getIdentifier() and large _BitInt

Fail gracefully when `getIdentifier()` for a declaration is null, such for
an overloaded operator.  Report an error and skip the declaration.

Fail gracefully when processing a large `_BitInt`, with a width greater
than 128.  Report an error and return `int` instead.

Remove the remaining fixed-point integral types from `convertType`.
Nothing can be done with them because fixed-point operations are not yet
implemented in the incubator, and I am not convinced that the types are
being converted correctly.

A couple small fixes to comments and documentation and formatting.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenTypes.cpp:
```diff
@@ -72,9 +63,15 @@ mlir::Type CIRGenTypes::convertType(QualType type) {
-                          bitIntTy->isSigned());
+    if (bitIntTy->getNumBits() > cir::IntType::maxBitwidth()) {
+      cgm.errorNYI(SourceLocation(), "large _BitInt type", type);
+      resultType = cir::IntType::get(cgm.getBuilder().getContext(), 32,
```
I might suggest we recover by doing IntType::maxBitwidth.  WDYT?

dkolsen-pgi replied:
Compilation has failed at this point, so it doesn't really matter exactly what type is used.  I like the consistency of always returning `int` for not-yet-implemented types.  (That code will become simpler when aliases for the common CIR types are created.)

erichkeane replied:
That seems like a reasonable take.  That seems consistent.

dkolsen-pgi commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.h:
```diff
@@ -0,0 +1,27 @@
+//===- CIRTypes.h - MLIR CIR Types ------------------------------*- C++ -*-===//
```
Done.

AaronBallman reviewed (approved):
LGTM
